tests/unit/modules/zypper_test.py

```diff
@@ -330,6 +330,76 @@ def test_latest_version(self):
+                ret = zypper.upgrade(dist_upgrade=True)
+                self.assertTrue(ret['result'])
+                self.assertDictEqual(ret['changes'], {"vim": {"old": "1.1", "new": "1.2"}})
+                zypper_mock.assert_any_call('dist-upgrade', '--auto-agree-with-licenses')
+
+            with patch('salt.modules.zypper.list_pkgs', MagicMock(side_effect=[{"vim": "1.1"}, {"vim": "1.1"}])):
+                ret = zypper.upgrade(dist_upgrade=True, dryrun=True)
+                self.assertTrue(ret['result'])
+                self.assertDictEqual(ret['changes'], {})
+                zypper_mock.assert_any_call('dist-upgrade', '--auto-agree-with-licenses', '--dry-run')
+                zypper_mock.assert_any_call('dist-upgrade', '--auto-agree-with-licenses', '--dry-run', '--debug-solver')
+
+            with patch('salt.modules.zypper.list_pkgs', MagicMock(side_effect=[{"vim": "1.1"}, {"vim": "1.2"}])):
+                ret = zypper.upgrade(dist_upgrade=True, fromrepo="Dummy", novendorchange=True)
+                self.assertTrue(ret['result'])
+                self.assertDictEqual(ret['changes'], {"vim": {"old": "1.1", "new": "1.2"}})
+                zypper_mock.assert_any_call('dist-upgrade', '--auto-agree-with-licenses', '--from', "Dummy", '--no-allow-vendor-change')
+
+    @patch('salt.modules.zypper.refresh_db', MagicMock(return_value=True))
+    @patch('salt.modules.zypper._systemd_scope', MagicMock(return_value=False))
+    def test_upgrade_failure(self):
+        '''
+        Test system upgrade failure.
+
+        :return:
+        '''
+        zypper_out = '''
+Loading repository data...
+Reading installed packages...
+Computing distribution upgrade...
+Use 'zypper repos' to get the list of defined repositories.
+Repository 'DUMMY' not found by its alias, number, or URI.
+'''
+
+        class FailingZypperDummy(object):
+            def __init__(self):
+                self.stdout = MagicMock(return_value=zypper_out)
+                self.stderr = MagicMock(return_value="")
+                self.exit_code = MagicMock(return_value=555)
+                self.noraise = MagicMock()
+                self.SUCCESS_EXIT_CODES = [0]
+
+            def __call__(self, *args, **kwargs):
+                return self
+
+        with patch('salt.modules.zypper.__zypper__', FailingZypperDummy()) as zypper_mock:
+            zypper_mock.noraise.call = MagicMock()
+            with patch('salt.modules.zypper.list_pkgs', MagicMock(side_effect=[{"vim": "1.1"}, {"vim": "1.1"}])):
+                    ret = zypper.upgrade(dist_upgrade=True, fromrepo="DUMMY")
+                    self.assertFalse(ret['result'])
+                    self.assertEqual(ret['comment'], zypper_out.strip())
+                    self.assertDictEqual(ret['changes'], {})
+                    zypper_mock.noraise.call.assert_called_with('dist-upgrade', '--auto-agree-with-licenses', '--from', 'DUMMY')
```

Comment:
Additionally to the previous things, I have no idea what had happened to your editor so the indentation went nuts. :-)